### ISSUE TITLE
update viewer app to pyqt5

### ISSUE THREAD

**Issue by thorade** [IfcOpenShell/IfcOpenShell#280]
The latest Anaconda comes with pyqt5 by default,
and to my understanding pythonocc prefers pyqt5 over pyqt4.

It is not possible to install pyqt4 and pyqt5 in parallel, one has to use virtual environments.

The viewer app currently uses pyqt4, and it would be awesome if it could be updated to pyqt5.
https://github.com/IfcOpenShell/IfcOpenShell/blob/master/src/ifcopenshell-python/ifcopenshell/geom/app.py
I do not know how easy or difficult that would be...

**Comment by aothms** [comment]
I agree. @jf--- has also brought this up in the past.

>  I do not know how easy or difficult that would be...

I can't imagine it to be too hard. But it will be a while until I will find the time. If you want to take the lead on this I can pitch in where needed.

**Comment by thorade** [comment]
Would it have to support pyqt4 and pyqt5, or is it ok to supprt pyqt5 ONLY?

**Comment by aothms** [comment]
> is it ok to supprt pyqt5 ONLY

I'd be fine with that. There's limited resources in a project like this so effort needs to be spent wisely.

http://blog.qt.io/blog/2014/11/27/qt-4-8-x-support-to-be-extended-for-another-year/

**Comment by thorade** [comment]
Here is a list of changes from pyqt4 to pyqt5:
http://pyqt.sourceforge.net/Docs/PyQt5/pyqt4_differences.html

Out of that list, I so far found two changes that are relevant:
 - QtGui has been split into QtGui and QtWidgets
    http://pyqt.sourceforge.net/Docs/PyQt5/pyqt4_differences.html#qtgui-module
    Whenever something is no longer found in QtGui, it has probably been moved to QtWidgets, so this is easy to fix.
 - pyqt5 no longer supports old-style signals:
    http://pyqt.sourceforge.net/Docs/PyQt5/pyqt4_differences.html#old-style-signals-and-slots
    This will be more work to fix.

**Comment by aothms** [comment]
Many thanks for looking into this.

> pyqt4 supports old-style and new-style signals and slots, pyqt5 only supports new-style:
> http://pyqt.sourceforge.net/Docs/PyQt4/new_style_signals_slots.html
> http://pyqt.sourceforge.net/Docs/PyQt5/pyqt4_differences.html#old-style-signals-and-slots
> This will be more work to change.

I find it a bit hard to follow that section. But from below, I don't think any of the following constructs are used. Are we already using 'new-style signals'?

> PyQt4’s old-style signals and slots are not supported. Therefore the following are not implemented in > PyQt5:
> 
>     QObject.connect()
>     QObject.emit()
>     SIGNAL()
>     SLOT()

**Comment by thorade** [comment]
No, unfortunately the current version uses old-style signals, and yes, I also did not really understand the docs. `.connect()` and `SIGNAL`are used in some places, and it seems there is no 1:1 translation.

**Comment by civilx64** [comment]
I had started some effort in this direction a few months ago but was very new to python and working outside of git.  @thorade, your outline of the work is consistent from what I remember - the API has been rearranged and the widgets inherit from a different hierarchy.  This was a relatively simple fix, but the change in slots and signals was baffling at the time.  

I took another dive into the qt API docs more recently and feel like I am starting to understand the new style events.  I'd be happy to help with this effort and can start a topic branch on my fork if one does not exist already.

**Comment by thorade** [comment]
I will not have time to work on this in the next weeks, so it would be great if you could work on it.
If you have a working pyqt4 setup, you could first port the old signals to new signals (as pyqt4 supports both). While that is the more difficult part, it would allow better intermediate testing.

Then importing/using the moved classes should be easy, but will break pyqt4 compatibility, and should be done in a second, seperate step.

**Comment by aothms** [comment]
> I'd be happy to help with this effort and can start a topic branch on my fork if one does not exist already.

That sound great! PR would be very welcome as this is something I'd have to read up on myself as well.

> but will break pyqt4 compatibility, and should be done in a second, seperate step.

Ideally in a separate commit as well so that we can review the potential pain of keeping pyqt4 support on the table (at least until a bump in version numbers).

Thanks!

**Comment by thorade** [comment]
Here is what the changes for the second part would look like:
https://github.com/thorade/IfcOpenShell/commit/c29faa4df4af01a4296c0c4fb3567bcf741c7315
But as said above, I would recommend starting with converting old signals into new signals.

**Comment by aothms** [comment]
> Here is what the changes for the second part would look like

Thanks! It seems reasonable to keep support qt4 for a while. With one or two `try catch`es and an alias for the module names we're almost there.

I will contribute the backwards compatibility :)

**Comment by civilx64** [comment]
I set up a virtualenv with pyqt4 and created a local branch for upgrading to new style signals and slots.  I'll post an update when I get stuck or have something ready to test.

**Comment by aothms** [comment]
Hey, are there any updates on this? Don't mean to appear pushy. Perhaps we can put what we have now on the table to see where we stand and what still needs to be done. With pythonocc et al on the anaconda ecosystem and well on pyqt5, the ifcopenshell pyqt4 app is close to useless now.

**Comment by thorade** [comment]
No updates from my side so far. I can try to have a look in the next few weeks.

**Comment by jf---** [comment]
You might want to consider using [`Qt.py`](https://github.com/mottosso/Qt.py);

> Qt.py enables you to write software that runs on any of the 4 supported bindings - PySide2, PyQt5, PySide and PyQt4.

**Comment by civilx64** [comment]
@aothms Nothing yet, sorry.  I had issues with the build and install, then appeared to be having hardware issues when I eventually did get the (PyQt4) viewer to run.  I was also running into the conda 'PackagesNotFoundError' like @johannesmichael in #318.  I just ran the install with explicit channel references and recommend adding those instructions to the readme.  I will send a PR on that soon.

Qt.py looks interesting.  I'm inclined to head in that direction, unless we want to avoid adding another dependency.  It appears to be installable via pip and conda, therefore should be easy to incorporate IMO.

**Comment by thorade** [comment]
Awesome! I pulled your branch, and it did also not work for me, but I ran the conversion tool again, and now I have a version that partly works here:
https://github.com/thorade/IfcOpenShell/tree/pyqt45_issue280
The viewer starts, but when I load an IFC file I get the error:
```
Traceback (most recent call last):
  File "C:\Users\mthorade\Anaconda3\lib\site-packages\ifcopenshell\geom\app.py", line 592, in browse
    self.load(filename)
  File "C:\Users\mthorade\Anaconda3\lib\site-packages\ifcopenshell\geom\app.py", line 605, in load
    c.load_file(f, setting=self.settings)
  File "C:\Users\mthorade\Anaconda3\lib\site-packages\ifcopenshell\geom\app.py", line 215, in load_file
    self.clicked[QModelIndex].connect(self.clicked)
NameError: name 'QModelIndex' is not defined
```

Also, the conversion might have others problems:
`inst.toPyObject()` seems to be converted incorrectly.
http://pyqt.sourceforge.net/Docs/PyQt5/pyqt_qvariant.html

Also, the conversion tool produces broken line breaks in some places.
I filed issues for all these.

**Comment by aothms** [comment]
Wonderful. It indeed didn't work out of the box, but was a good starting point. I committed some stuff. It's still not completely back the way it should, selection from the treeview does not work, either due to the missing `QModelIndex` or this `QVarient` stuff brought up by @thorade. But at least there is something on the screen again.

**Comment by Moult** [comment]
Fast forward to Qt6 :)
